File: src/server/domain/sourceAuthority.ts
/**
 * Which exact stored `Source` a request named — one authority, shared by every door.
 *
 * IR-107 B2-C. This function lived inside `askMarket`, where it served the deterministic path and
 * nothing else. That was the whole problem, and it is not a tidiness complaint: measured against
 * real PostgreSQL, with only provider Y publishing a subject, a request naming provider X came back
 * AUTHORIZED carrying Y's series on BOTH the canonical and legacy candidate envelopes, and the
 * planner was called with it. The deterministic path refused the same request correctly, because it
 * was the only path that could see this code.
 *
 * A wrong-source answer is the worst shape of error this product has. The figure is real, the
 * subject is right, and the attribution is false — which is exactly what makes it credible.
 *
 * ## Grammar and identity stay apart
 *
 * `RequestAuthority.sourceRegion` carries the source constituent as TEXT: the grammar proved
 * something occupied the source role, and proved nothing about whether this repository holds such a
 * source. Identity is established here, against stored rows. A caller or a model asserting a
 * `sourceId` is never a substitute, the parser must not resolve identity, and this module must not
 * parse.
 *
 * ## Two matches is not a tie
 *
 * `Source.name` is free text and is not unique, so two rows may legitimately answer to the same
 * name, and a request in which two different providers both occur did not say which one it meant.
 * AMBIGUOUS is the answer — not the longer one, and not the first row.
 *
 * ## What was NOT promoted
 *
 * An untracked draft of this module existed in the tree, named by two planning documents. It was
 * the PRE-COVER version of the same function: no full-role cover, no `RESIDUE` status. Staging it
 * because a document mentioned it would have reintroduced the defect ESC-015 §8 closed. The tracked
 * logic was moved here instead and the draft overwritten.
 */

import { prisma } from "@/server/db/client";
import { explicitlyNamed, regionIsExactlyFramingAndIdentity } from "./subjectAuthority";
import { requestFramingIsRecognised } from "./requestAuthority";
import { isGenericThirdPartyReference } from "./requestFrame";

/**
 * A named source resolved to a repository identity, or a refusal.
 *
 * The parsed source constituent is TEXT. It becomes authority only by matching a `Source` this
 * repository actually holds, and matching more than one is not a tie to be broken — two providers
 * whose names both occur in the request means the request did not say which.
 */
export type SourceResolution =
  | { status: "RESOLVED"; sourceId: string; code: string }
  | { status: "AMBIGUOUS"; codes: string[] }
  | { status: "RESIDUE" }
  | { status: "UNRESOLVED" };

export async function resolveSourceIdentity(sourceRegion: string): Promise<SourceResolution> {
  const region = sourceRegion.trim();
  if (!region) return { status: "UNRESOLVED" };
  const sources = await prisma.source.findMany({
    select: { id: true, code: true, name: true },
    // A unique tiebreak, per `orderingDeterminism`: source codes are unique today, and an
    // ordering that relies on that staying true is an ordering that can tie tomorrow.
    orderBy: [{ code: "asc" }, { id: "asc" }],
  });
  // Containment of the WHOLE name, not overlap. `mentionsEachOther` is a retrieval heuristic and
  // it reported both "Test PB Source A" and "Test PB Source B" as matching a request that named
  // one of them -- three shared words out of four. Retrieval may guess; identity may not.
  //
  // Unicode-aware, matching `subjectAuthority.normalizeSubject`. An ASCII-only character class
  // erases a non-Latin source name to the empty string, and the empty string is contained in
  // every request -- so the check would have resolved every source, or none, on a name it could
  // not see.
  const normalize = (text: string) =>
    ` ${text
      .normalize("NFKC")
      .toLowerCase()
      .replace(/[^\p{L}\p{N}]+/gu, " ")
      .trim()} `;
  const haystack = normalize(region);
  // A code is not a name and does not nest the way names do, so it is matched separately.
  const byCode = sources.filter((src) => haystack.includes(normalize(src.code)));
  // Keep only maximal names -- by OCCURRENCE, for the same reason as subjects. Naming the longer
  // source makes both whole names occur and the shorter was read out of the longer, so it must go;
  // but naming BOTH ("Rework Data, Rework Data Research") names two providers, and a name-level
  // containment test could not tell those two situations apart. It answered the second one by
  // silently picking the longer provider, which is a false attribution with a confident tone.
  // Same deletion as the series path: the whole-name containment pre-filter decided nothing that
  // `explicitlyNamed` does not decide again, since a name with no occurrence has no occurrence.
  const byName = explicitlyNamed(sources, (src) => src.name, region);
  const hits = [...new Map([...byName, ...byCode].map((src) => [src.id, src])).values()];

  // FULL-ROLE COVER on the source role. ESC-015 §8.
  //
  // Occurrence found which providers are worth considering. It does not settle whether the role
  // said anything else, and until this ran, it did not have to: measured against a real repository,
  // `What did <provider> Purchase Gamma shares publish about <series>?` returned FACTORS_FOUND and
  // published the reading under that provider's attribution. The parser hands this function
  // `<provider> purchase gamma shares` as the source region and considers the request authorized --
  // `scripts/probe-source-role-residue.ts` prints five such regions.
  //
  // Attribution is a stronger claim than a figure. A reading published as "what <provider>
  // reported" says a named organisation said something, so the role that names it has to be
  // explainable by that name and framing, with nothing left over.
  //
  // Either the name or the code may be the identity that explains the role, because both are ways
  // a request can name a provider and `byCode` above matches on the second. The rule itself is the
  // one the subject and relation roles use; only the vocabulary is chosen here.
  const covering = hits.filter(
    (src) =>
      regionIsExactlyFramingAndIdentity(region, src.name, requestFramingIsRecognised) ||
      regionIsExactlyFramingAndIdentity(region, src.code, requestFramingIsRecognised),
  );
  if (covering.length === 1) {
    return { status: "RESOLVED", sourceId: covering[0].id, code: covering[0].code };
  }
  if (covering.length > 1) return { status: "AMBIGUOUS", codes: covering.map((h) => h.code) };
  // A provider was named and the role said more. That is not the same as naming no provider this
  // repository holds, and the caller reports them differently -- one is a role this grammar could
  // not account for, the other is a genuine gap in inventory.
  if (hits.length > 0) return { status: "RESIDUE" };
  return { status: "UNRESOLVED" };
}

/**
 * Did the source role name a PARTY at all, or only say that somebody else reported it?
 *
 * The B2-C repair binds an attributed request to exactly one stored provider, and applied without
 * this distinction it removes the operation rather than securing it: every frame-eligible attributed
 * request in this repository's corpus reads `What did analysts publish about X?`, and `analysts` is
 * the vocabulary that PROVES the third-party frame, not a party. Requiring it to resolve to a stored
 * `Source` refused 54 tests' worth of ordinary behaviour.
 *
 * Wrong-source substitution is publishing Y's record for a request that named X. A request naming
 * nobody cannot suffer it. What it must not do is name a provider this repository does not hold and
 * be answered from one it does — which is the other half, and stays refused.
 */
export function sourceRoleNamesAParty(sourceRegion: string): boolean {
  return sourceRegion.trim() !== "" && !isGenericThirdPartyReference(sourceRegion);
}

File: scripts/mutation/sourceauthority.py
"""M-SRCAUTH: is exact source authority load-bearing on the planner-facing doors?

WRITE/EDIT TOOL ONLY -- heredocs in this environment eat backslashes, and these anchors carry
escaped newlines. Five files in this directory have been corrupted that way.

IR-107 B2-C. Before this unit, attribution was not a candidate-authority dimension at all: the
canonical envelope resolved ATTRIBUTED_REPORTED_OBSERVATION by subject identity alone and never read
`request.sourceRegion`, and the legacy envelope delegated to subject authority the same way. With
only provider Y publishing a subject, a request naming provider X came back AUTHORIZED carrying Y's
series on both doors, and `answerWithInference` called the planner with it.

Four mutants must turn RED. The fifth must NOT.

  M-SRCAUTH-ORDER is inverted on purpose: it reverses the order stored providers come back in, and
  the REQUIRED result is MISSED. Which provider answers must be decided by the request, never by
  enumeration. If it is ever reported ISOLATED, some test has begun depending on row order and that
  is the finding, not a pass.

    python scripts/mutation/sourceauthority.py [ID ...]
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from harness import harness

ENVELOPE = "src/server/domain/candidateEnvelope.ts"
SOURCE = "src/server/domain/sourceAuthority.ts"

BINDING_TESTS = [
    "tests/integration/source-authority.test.ts",
    # The doors' pre-existing behaviour, which this unit must not have bought its guarantee with.
    "tests/integration/canonical-candidate.test.ts",
    "tests/integration/source-role-cover.test.ts",
]
UNRELATED_TESTS = [
    # Roles with no source dimension at all.
    "tests/integration/relation-role-cover.test.ts",
    "tests/framingPositionalAuthority.test.ts",
]

MUTATIONS = [
    # M-SRCAUTH-FILTER -- the candidate series stop being narrowed to the named provider. This is
    # the defect exactly as reproduced: the subject is resolved across every provider, so Y's row
    # answers a request naming X.
    (
        "M-SRCAUTH-FILTER candidates are no longer narrowed to the named provider",
        ENVELOPE,
        "      const owned = await prisma.series.findMany({\n"
        "        where: { sourceId: source.sourceId },\n"
        "        select: { id: true, name: true },\n"
        "      });",
        "      const owned = await prisma.series.findMany({ select: { id: true, name: true } });",
    ),
    # M-SRCAUTH-OCCURRENCE -- exact source cover degrades to occurrence. A provider name occurring
    # anywhere inside a longer source role authorizes attribution to it again, which is the ESC-015
    # §8 defect arriving through the planner door instead of the deterministic one.
    (
        "M-SRCAUTH-OCCURRENCE exact source cover degrades to occurrence",
        SOURCE,
        "  const covering = hits.filter(\n"
        "    (src) =>\n"
        "      regionIsExactlyFramingAndIdentity(region, src.name, requestFramingIsRecognised) ||\n"
        "      regionIsExactlyFramingAndIdentity(region, src.code, requestFramingIsRecognised),\n"
        "  );",
        "  const covering = hits;",
    ),
    # M-SRCAUTH-FIRST -- two providers answering to one name are resolved by taking the first.
    # `Source.name` is free text and is not unique, so this is reachable with ordinary data.
    (
        "M-SRCAUTH-FIRST an ambiguous source is resolved by taking the first",
        SOURCE,
        '  if (covering.length > 1) return { status: "AMBIGUOUS", codes: covering.map((h) => h.code) };',
        "  if (covering.length > 1) {\n"
        '    return { status: "RESOLVED", sourceId: covering[0].id, code: covering[0].code };\n'
        "  }",
    ),
    # M-SRCAUTH-LEGACY-FRAME -- the legacy door goes back to treating frame eligibility as proof of
    # a source. Separate from the canonical mutants because the two doors are reached by different
    # requests, and a repair that only landed on one of them would survive the others.
    (
        "M-SRCAUTH-LEGACY-FRAME frame eligibility is proof of a source again",
        ENVELOPE,
        '  if (authority.status === "AUTHORIZED" && authority.operation === "REPORTED_OBSERVATION") {',
        "  if (false) {",
    ),
    # M-SRCAUTH-ORDER -- REQUIRED TO SURVIVE.
    (
        "M-SRCAUTH-ORDER stored providers are enumerated in the opposite order",
        SOURCE,
        '    orderBy: [{ code: "asc" }, { id: "asc" }],',
        '    orderBy: [{ code: "desc" }, { id: "desc" }],',
    ),
]

SELECTED = sys.argv[1:]
if SELECTED:
    MUTATIONS = [m for m in MUTATIONS if any(m[0].startswith(s) for s in SELECTED)]
    if not MUTATIONS:
        print(f"no mutant matches {SELECTED}")
        sys.exit(3)
    print(f"PARTIAL RUN: {len(MUTATIONS)} of 5. Not a substitute for the full set.")

sys.exit(harness([ENVELOPE, SOURCE], BINDING_TESTS, UNRELATED_TESTS, MUTATIONS, wall_seconds=2400))
